src/compiler/checker.ts

```diff
@@ -5182,7 +5220,14 @@ namespace ts {
         }
 
         function isContextSensitiveFunctionLikeDeclaration(node: FunctionLikeDeclaration) {
-            return !node.typeParameters && node.parameters.length && !forEach(node.parameters, p => p.type);
+            if (compilerOptions.strictThis) {
+                return !node.typeParameters &&
+                    (!forEach(node.parameters, p => p.type)

```

Comment:
Maybe add a `hasParameterWithTypeAnnotation` function and do the appropriate compensation for `this` parameters in there.